### PR TITLE
V2 java doc

### PR DESCRIPTION
**Pull request by MechaRyRy**
## Problem
We are getting closer and closer to performing the major release of `download-manager` v2 and we have very little in the way of javadocs.

## Solution
Begin to add some javadocs.

### CODE REVIEW

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: I prefer `Defines` instead of `For defining`

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: `Clients should control the throttling of this status. and sequently notify the callback defined with setCallback()` with a link to the method, I never remember how to do it 😅

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: This is not true, this is more like "notify the callback that no more DownloadBatchStatus will be emitted". Should we also specify that the implementer of this class might want notify the final callback with the last emitted status?

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java:8]: While we are at it, should we rename this?
It still sounds super weird to me... at least removing the `lite` part?

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: I don't remember, how can the client use this?

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java]: Maybe add an example of when this should be called? like after migration completed, or after reboot (if needed)...

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java:73]: `Asynchronously retrieves a list of {@link DownloadBatchStatus} passing it to the {@link AllBatchStatusesCallback}.`

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java:93]: Asynchronously retrieves

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java]: ...and resume paused ones once the condition is satisfied

File: library/src/main/java/com/novoda/downloadmanager/NotificationCustomizer.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/NotificationCustomizer.java]: Specify that this should return the display state (with a link to the class) for the given payload, so that the user can decide if to show/hide the notification.
Also, missing javadoc for that class

File: library/src/main/java/com/novoda/downloadmanager/NotificationCustomizer.java
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/NotificationCustomizer.java:14]: This class should have a description and link to `DownloadManagerBuilder` method using it

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (danybony) [reply library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: Ah yes, this class should have a link to `DownloadManagerBuilder` method using it

**Review by danybony — CHANGES_REQUESTED** [review]

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java:73]: I would avoid saying that something does something asynchronously. We have the `WorkerThread` annotation to show when something is done synchronously, seems overkill to then say the same for async. What do you think @zegnus ?

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: Good idea! 💯

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java:8]: Yea I might do that actually, but it might be better left until the migration stuff is merged, this will have lots of conflicts probably 😄

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: I don't follow. This method prevents additional statuses from being emitted. Either with a guard statement or by removing a listener. It doesn't notify any callback to stop doing something.

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: Good idea 😄

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: 👍

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (danybony) [reply library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java:8]: Agreed 👍

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java]: I was confused about this because I couldn't see where it resumes once the condition is satisfied. It doesn't seem to schedule a job or anything to check.

File: library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java
Comment (danybony) [reply library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: What I meant is that given how it's phrased like now, as a client I'm implementing this interface with my concrete implementation. This method is gonna then be called to tell my implementation that no more batch statuses are gonna be emitted. It's a callback, not a method that actually prevents the emission (that I need to implemented myself). I don't know if it's clearer now

Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/CallbackThrottle.java]: Oh I see what you mean. It's a method that is called by the `download-manager` to prevent the statuses from being emitted, as opposed to something that the client can call to stop them 👍 makes sense 😄

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java]: f0bbd450ec7e8c630f0006fce1a4ebf67612762b

File: library/src/main/AndroidManifest.xml
Comment (zegnus) [inline library/src/main/AndroidManifest.xml:8]: why is this part of a javadoc PR?

File: demo-simple/src/main/res/xml/network_security_config.xml
Comment (MechaRyRy) [inline demo-simple/src/main/res/xml/network_security_config.xml:3]: Starting with Android 9.0 (API level 28), cleartext support is disabled by default. You need to specify this file or do a hack on the application in the manifest to workaround it
The idea being that the app should specify what endpoints it can interact with rather than just connecting to anything :joy:
I specify `true` so that it behaves as it did prior to API 28 and have added it to the demo to show how people can override it

File: library/src/main/java/com/novoda/downloadmanager/DownloadBatchStatus.java
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/DownloadBatchStatus.java:44]: Maybe we can add that this title is going to be used by the default notifications

File: library/src/main/java/com/novoda/downloadmanager/DownloadFileStatus.java
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/DownloadFileStatus.java:42]: Worth adding an example of the format?

File: library/src/main/java/com/novoda/downloadmanager/FilePath.java
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/FilePath.java]: same as before, it would be interesting to know if this returns a relative or an absolute path for example

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java]: >repopulating the internal map
might not be necessary to say

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (zegnus) [reply library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java:73]: Agree with @Mecharyry, by default calls are non-blocking and if the param is a callback I would understand that asynchronicity. Blocking calls though should be noted in the Javadoc.

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java]: I would add something like, This is a long-running blocking operation, clients should call this method on their own Threading mechanism. Same for everywhere we use synchronous.

I'm mentioning this because that a method is synchronous is not a problem, the majority are. The problem is that this is a long running blocking operation.

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java]: when might this method be useful for a client?

File: library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java
Comment (danybony) [reply library/src/main/java/com/novoda/downloadmanager/LiteDownloadManagerCommands.java:73]: I still think that we should add the Synchronously/Asynchronously words when there are two variations of the same method

**Review by danybony — COMMENTED** [review]
Just one comment about sync/async to be added to some methods description, up to @zegnus

**Review by danybony — APPROVED** [review]

File: library/src/main/AndroidManifest.xml
Comment (MechaRyRy) [reply library/src/main/AndroidManifest.xml:8]: 🤔 not sure. I think I updated to API 28 when the build was failing 😬

**Comment by MechaRyRy** [comment]
@zegnus I've been through and addressed all of your comments.

**Review by zegnus — APPROVED** [review]

**Pull request merged**
